# 🔧 Fase 3: Preparación de los Datos NBA

## 📋 Objetivo
Transformar los datos de la NBA para que estén listos para el modelado de machine learning, aplicando técnicas de limpieza, transformación y división basadas en los insights obtenidos en las fases anteriores.

## 🎯 Metodología
Esta fase se enfoca en:
- **Limpieza sistemática** de datos basada en el análisis de la Fase 2
- **Transformaciones inteligentes** que preserven la información relevante
- **División estratégica** del dataset para validación robusta
- **Justificación técnica** de cada decisión tomada

---

## 📊 Contenido del Análisis
1. **Carga y Revisión** - Recuperar insights de fases anteriores
2. **Limpieza de Datos** - Imputación, outliers, inconsistencias
3. **Transformaciones** - Codificación, normalización, fechas
4. **División Estratégica** - Train/test split con validación
5. **Justificación Técnica** - Fundamentos estadísticos y matemáticos
6. **Dataset Final** - Verificación y documentación

---

## 🧠 Fundamentos Teóricos
- **Estadística Descriptiva**: Medidas de tendencia central y dispersión
- **Álgebra Lineal**: Transformaciones matriciales y escalado
- **Teoría de Probabilidad**: Distribuciones y muestreo
- **Machine Learning**: Preprocesamiento y validación cruzada


In [1]:
# 📥 Carga de Datos y Revisión de Insights
print("🔄 CARGANDO DATOS Y REVISANDO INSIGHTS")
print("=" * 50)

# Importar librerías necesarias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder, OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

# Configurar estilo de visualizaciones
plt.style.use('seaborn-v0_8')
sns.set_palette("husl")

# Cargar dataset principal
print("📊 Cargando dataset principal...")
games_df = pd.read_csv('../data/01_raw/game.csv')
print(f"✅ Dataset cargado: {games_df.shape}")

# Cargar estadísticas adicionales para enriquecimiento
other_stats_df = pd.read_csv('../data/01_raw/other_stats.csv')
print(f"✅ Estadísticas adicionales cargadas: {other_stats_df.shape}")

# Revisar insights de fases anteriores
print(f"\n🔍 REVISIÓN DE INSIGHTS DE FASES ANTERIORES:")
print(f"• Total de partidos: {len(games_df):,}")
print(f"• Variables disponibles: {len(games_df.columns)}")
print(f"• Rango temporal: {games_df['game_date'].min()} a {games_df['game_date'].max()}")
print(f"• Temporadas: {games_df['season_id'].nunique()}")

# Verificar variable objetivo
win_percentage = (games_df['wl_home'] == 'W').mean()
print(f"• Porcentaje de victorias locales: {win_percentage:.1%}")

# Identificar tipos de variables
numeric_cols = games_df.select_dtypes(include=[np.number]).columns.tolist()
categorical_cols = games_df.select_dtypes(include=['object']).columns.tolist()
print(f"• Variables numéricas: {len(numeric_cols)}")
print(f"• Variables categóricas: {len(categorical_cols)}")

print("\n✅ Carga y revisión completada")


🔄 CARGANDO DATOS Y REVISANDO INSIGHTS
📊 Cargando dataset principal...
✅ Dataset cargado: (65698, 55)
✅ Estadísticas adicionales cargadas: (28271, 26)

🔍 REVISIÓN DE INSIGHTS DE FASES ANTERIORES:
• Total de partidos: 65,698
• Variables disponibles: 55
• Rango temporal: 1946-11-01 00:00:00 a 2023-06-12 00:00:00
• Temporadas: 225
• Porcentaje de victorias locales: 61.9%
• Variables numéricas: 45
• Variables categóricas: 10

✅ Carga y revisión completada


In [2]:
# 🧹 Limpieza de Datos
print("🧹 LIMPIEZA DE DATOS")
print("=" * 30)

# Crear copia del dataset para trabajar
df_clean = games_df.copy()
print(f"📊 Dataset original: {df_clean.shape}")

# 1. ANÁLISIS DE VALORES NULOS
print(f"\n❌ 1. ANÁLISIS DE VALORES NULOS")
print("-" * 40)

# Calcular valores nulos por columna
null_analysis = pd.DataFrame({
    'Valores_Nulos': df_clean.isnull().sum(),
    'Porcentaje_Nulos': (df_clean.isnull().sum() / len(df_clean)) * 100
}).sort_values('Porcentaje_Nulos', ascending=False)

# Filtrar columnas con valores nulos
null_columns = null_analysis[null_analysis['Valores_Nulos'] > 0]
print(f"📊 Columnas con valores nulos: {len(null_columns)}")

if len(null_columns) > 0:
    print("🔍 Top 10 columnas con más valores nulos:")
    display(null_columns.head(10))
    
    # Estrategia de imputación basada en el tipo de variable
    print(f"\n🎯 ESTRATEGIA DE IMPUTACIÓN:")
    
    # Para variables numéricas: usar mediana (robusta a outliers)
    numeric_null_cols = [col for col in null_columns.index if col in numeric_cols]
    if numeric_null_cols:
        print(f"• Variables numéricas ({len(numeric_null_cols)}): Imputación con mediana")
        for col in numeric_null_cols:
            median_value = df_clean[col].median()
            df_clean[col].fillna(median_value, inplace=True)
            print(f"  - {col}: {df_clean[col].isnull().sum()} → 0 (mediana: {median_value:.2f})")
    
    # Para variables categóricas: usar moda
    categorical_null_cols = [col for col in null_columns.index if col in categorical_cols]
    if categorical_null_cols:
        print(f"• Variables categóricas ({len(categorical_null_cols)}): Imputación con moda")
        for col in categorical_null_cols:
            mode_value = df_clean[col].mode().iloc[0] if not df_clean[col].mode().empty else 'Unknown'
            df_clean[col].fillna(mode_value, inplace=True)
            print(f"  - {col}: {df_clean[col].isnull().sum()} → 0 (moda: {mode_value})")

# Verificar que no queden valores nulos
remaining_nulls = df_clean.isnull().sum().sum()
print(f"\n✅ Valores nulos restantes: {remaining_nulls}")

# 2. DETECCIÓN Y TRATAMIENTO DE OUTLIERS
print(f"\n🔍 2. DETECCIÓN Y TRATAMIENTO DE OUTLIERS")
print("-" * 50)

def detect_outliers_iqr(data, column):
    """Detectar outliers usando el método IQR (Rango Intercuartílico)"""
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = data[(data[column] < lower_bound) | (data[column] > upper_bound)]
    return outliers, lower_bound, upper_bound

# Variables clave para análisis de outliers
outlier_vars = ['pts_home', 'pts_away', 'fg_pct_home', 'fg_pct_away', 
                'reb_home', 'reb_away', 'ast_home', 'ast_away']

outlier_summary = []
outliers_removed = 0

print("📊 Análisis de outliers (Método IQR):")
for var in outlier_vars:
    if var in df_clean.columns:
        outliers, lower, upper = detect_outliers_iqr(df_clean, var)
        outlier_count = len(outliers)
        outlier_percentage = (outlier_count / len(df_clean)) * 100
        
        outlier_summary.append({
            'Variable': var,
            'Outliers': outlier_count,
            'Porcentaje': f"{outlier_percentage:.2f}%",
            'Límite_Inferior': f"{lower:.2f}",
            'Límite_Superior': f"{upper:.2f}"
        })
        
        # Estrategia: Cap outliers en lugar de eliminarlos (preservar información)
        if outlier_count > 0:
            df_clean[var] = np.clip(df_clean[var], lower, upper)
            outliers_removed += outlier_count
            print(f"  • {var}: {outlier_count} outliers capados ({outlier_percentage:.2f}%)")

print(f"\n✅ Total de outliers tratados: {outliers_removed}")

# 3. CORRECCIÓN DE INCONSISTENCIAS
print(f"\n🔧 3. CORRECCIÓN DE INCONSISTENCIAS")
print("-" * 40)

# Verificar consistencia en fechas
df_clean['game_date'] = pd.to_datetime(df_clean['game_date'])
print(f"✅ Fechas convertidas a datetime")

# Verificar consistencia en porcentajes (deben estar entre 0 y 1)
percentage_cols = [col for col in df_clean.columns if 'pct' in col.lower()]
for col in percentage_cols:
    if col in df_clean.columns:
        # Convertir porcentajes que estén en escala 0-100 a 0-1
        if df_clean[col].max() > 1:
            df_clean[col] = df_clean[col] / 100
            print(f"  • {col}: Convertido de escala 0-100 a 0-1")

# Verificar que no haya valores negativos en estadísticas que no deberían tenerlos
positive_cols = ['pts_home', 'pts_away', 'reb_home', 'reb_away', 'ast_home', 'ast_away']
for col in positive_cols:
    if col in df_clean.columns:
        negative_count = (df_clean[col] < 0).sum()
        if negative_count > 0:
            df_clean[col] = np.maximum(df_clean[col], 0)
            print(f"  • {col}: {negative_count} valores negativos corregidos a 0")

print(f"\n📊 Dataset después de limpieza: {df_clean.shape}")
print("✅ Limpieza de datos completada")


🧹 LIMPIEZA DE DATOS
📊 Dataset original: (65698, 55)

❌ 1. ANÁLISIS DE VALORES NULOS
----------------------------------------
📊 Columnas con valores nulos: 36
🔍 Top 10 columnas con más valores nulos:


Valores_Nulos  Porcentaje_Nulos
fg3_pct_home          19074         29.032847
dreb_home             18999         28.918689
dreb_away             18998         28.917166
fg3_pct_away          18962         28.862370
oreb_away             18936         28.822795
oreb_home             18936         28.822795
stl_home              18849         28.690371
stl_away              18849         28.690371
tov_away              18685         28.440744
tov_home              18684         28.439222


🎯 ESTRATEGIA DE IMPUTACIÓN:
• Variables numéricas (34): Imputación con mediana
  - fg3_pct_home: 0 → 0 (mediana: 0.35)
  - dreb_home: 0 → 0 (mediana: 31.00)
  - dreb_away: 0 → 0 (mediana: 30.00)
  - fg3_pct_away: 0 → 0 (mediana: 0.33)
  - oreb_away: 0 → 0 (mediana: 11.00)
  - oreb_home: 0 → 0 (mediana: 12.00)
  - stl_home: 0 → 0 (mediana: 8.00)
  - stl_away: 0 → 0 (mediana: 8.00)
  - tov_away: 0 → 0 (mediana: 15.00)
  - tov_home: 0 → 0 (mediana: 15.00)
  - fg3a_away: 0 → 0 (mediana: 16.00)
  - fg3a_home: 0 → 0 (mediana: 16.00)
  - blk_home: 0 → 0 (mediana: 5.00)
  - blk_away: 0 → 0 (mediana: 4.00)
  - ast_home: 0 → 0 (mediana: 24.00)
  - ast_away: 0 → 0 (mediana: 22.00)
  - reb_home: 0 → 0 (mediana: 43.00)
  - reb_away: 0 → 0 (mediana: 42.00)
  - fg_pct_home: 0 → 0 (mediana: 0.47)
  - fg_pct_away: 0 → 0 (mediana: 0.46)
  - fga_home: 0 → 0 (mediana: 84.00)
  - fga_away: 0 → 0 (mediana: 83.00)
  - fg3m_home: 0 → 0 (mediana: 5.00)
  - fg3m_away: 0 → 0 (mediana: 5.00)
  - ft_pct_home: 0 →

In [3]:
# 🔄 Transformaciones de Datos
print("🔄 TRANSFORMACIONES DE DATOS")
print("=" * 35)

# Crear copia para transformaciones
df_transformed = df_clean.copy()

# 1. CONVERSIÓN DE FECHAS A VARIABLES ÚTILES
print(f"\n📅 1. CONVERSIÓN DE FECHAS A VARIABLES ÚTILES")
print("-" * 50)

# Extraer componentes de fecha
df_transformed['year'] = df_transformed['game_date'].dt.year
df_transformed['month'] = df_transformed['game_date'].dt.month
df_transformed['day'] = df_transformed['game_date'].dt.day
df_transformed['day_of_week'] = df_transformed['game_date'].dt.dayofweek  # 0=Lunes, 6=Domingo
df_transformed['day_of_year'] = df_transformed['game_date'].dt.dayofyear
df_transformed['week_of_year'] = df_transformed['game_date'].dt.isocalendar().week

# Crear variables estacionales
df_transformed['is_weekend'] = (df_transformed['day_of_week'] >= 5).astype(int)
df_transformed['is_playoff_season'] = (df_transformed['month'].isin([4, 5, 6])).astype(int)  # Abril-Junio

print("✅ Variables de fecha creadas:")
print("  • year, month, day, day_of_week, day_of_year, week_of_year")
print("  • is_weekend, is_playoff_season")

# 2. CODIFICACIÓN DE VARIABLES CATEGÓRICAS
print(f"\n🏷️ 2. CODIFICACIÓN DE VARIABLES CATEGÓRICAS")
print("-" * 50)

# Identificar variables categóricas relevantes
categorical_features = ['wl_home', 'season_type', 'team_abbreviation_home', 'team_abbreviation_away']

# Crear variable objetivo binaria
df_transformed['home_win'] = (df_transformed['wl_home'] == 'W').astype(int)
print(f"✅ Variable objetivo binaria creada: home_win")

# Codificar season_type (One-Hot Encoding para preservar información)
season_type_dummies = pd.get_dummies(df_transformed['season_type'], prefix='season')
df_transformed = pd.concat([df_transformed, season_type_dummies], axis=1)
print(f"✅ season_type codificado con One-Hot: {list(season_type_dummies.columns)}")

# Codificar equipos (Label Encoding para reducir dimensionalidad)
le_home = LabelEncoder()
le_away = LabelEncoder()

df_transformed['team_home_encoded'] = le_home.fit_transform(df_transformed['team_abbreviation_home'])
df_transformed['team_away_encoded'] = le_away.fit_transform(df_transformed['team_abbreviation_away'])
print(f"✅ Equipos codificados con Label Encoding")
print(f"  • Equipos locales únicos: {df_transformed['team_home_encoded'].nunique()}")
print(f"  • Equipos visitantes únicos: {df_transformed['team_away_encoded'].nunique()}")

# 3. CREACIÓN DE VARIABLES DERIVADAS
print(f"\n🧮 3. CREACIÓN DE VARIABLES DERIVADAS")
print("-" * 40)

# Diferenciales entre equipos (más informativos que valores absolutos)
df_transformed['pts_diff'] = df_transformed['pts_home'] - df_transformed['pts_away']
df_transformed['fg_pct_diff'] = df_transformed['fg_pct_home'] - df_transformed['fg_pct_away']
df_transformed['reb_diff'] = df_transformed['reb_home'] - df_transformed['reb_away']
df_transformed['ast_diff'] = df_transformed['ast_home'] - df_transformed['ast_away']
df_transformed['stl_diff'] = df_transformed['stl_home'] - df_transformed['stl_away']
df_transformed['blk_diff'] = df_transformed['blk_home'] - df_transformed['blk_away']
df_transformed['tov_diff'] = df_transformed['tov_home'] - df_transformed['tov_away']

print("✅ Variables diferenciales creadas:")
print("  • pts_diff, fg_pct_diff, reb_diff, ast_diff, stl_diff, blk_diff, tov_diff")

# Ratios de eficiencia
df_transformed['home_efficiency'] = df_transformed['pts_home'] / (df_transformed['fga_home'] + 0.001)  # Evitar división por 0
df_transformed['away_efficiency'] = df_transformed['pts_away'] / (df_transformed['fga_away'] + 0.001)
df_transformed['efficiency_diff'] = df_transformed['home_efficiency'] - df_transformed['away_efficiency']

print("✅ Variables de eficiencia creadas:")
print("  • home_efficiency, away_efficiency, efficiency_diff")

# 4. NORMALIZACIÓN Y ESTANDARIZACIÓN
print(f"\n📏 4. NORMALIZACIÓN Y ESTANDARIZACIÓN")
print("-" * 45)

# Seleccionar variables numéricas para escalado
numeric_features = [
    'pts_home', 'pts_away', 'fg_pct_home', 'fg_pct_away', 'fg3_pct_home', 'fg3_pct_away',
    'ft_pct_home', 'ft_pct_away', 'reb_home', 'reb_away', 'oreb_home', 'oreb_away',
    'dreb_home', 'dreb_away', 'ast_home', 'ast_away', 'stl_home', 'stl_away',
    'blk_home', 'blk_away', 'tov_home', 'tov_away', 'pf_home', 'pf_away',
    'plus_minus_home', 'plus_minus_away', 'pts_diff', 'fg_pct_diff', 'reb_diff',
    'ast_diff', 'stl_diff', 'blk_diff', 'tov_diff', 'efficiency_diff'
]

# Filtrar variables que existen en el dataset
available_numeric_features = [col for col in numeric_features if col in df_transformed.columns]

print(f"📊 Variables seleccionadas para escalado: {len(available_numeric_features)}")

# Aplicar StandardScaler (media=0, desv_std=1)
scaler = StandardScaler()
df_transformed[available_numeric_features] = scaler.fit_transform(df_transformed[available_numeric_features])

print("✅ StandardScaler aplicado:")
print(f"  • Media de variables escaladas: {df_transformed[available_numeric_features].mean().mean():.6f}")
print(f"  • Desviación estándar: {df_transformed[available_numeric_features].std().mean():.6f}")

# 5. VERIFICACIÓN DE TRANSFORMACIONES
print(f"\n✅ 5. VERIFICACIÓN DE TRANSFORMACIONES")
print("-" * 40)

print(f"📊 Dataset después de transformaciones: {df_transformed.shape}")
print(f"• Variables originales: {len(games_df.columns)}")
print(f"• Variables después de transformaciones: {len(df_transformed.columns)}")
print(f"• Nuevas variables creadas: {len(df_transformed.columns) - len(games_df.columns)}")

# Verificar que no hay valores infinitos o NaN
inf_count = np.isinf(df_transformed.select_dtypes(include=[np.number])).sum().sum()
nan_count = df_transformed.isnull().sum().sum()

print(f"• Valores infinitos: {inf_count}")
print(f"• Valores nulos: {nan_count}")

if inf_count > 0:
    print("⚠️  Corrigiendo valores infinitos...")
    df_transformed = df_transformed.replace([np.inf, -np.inf], np.nan)
    df_transformed = df_transformed.fillna(0)

print("✅ Transformaciones completadas")


🔄 TRANSFORMACIONES DE DATOS

📅 1. CONVERSIÓN DE FECHAS A VARIABLES ÚTILES
--------------------------------------------------
✅ Variables de fecha creadas:
  • year, month, day, day_of_week, day_of_year, week_of_year
  • is_weekend, is_playoff_season

🏷️ 2. CODIFICACIÓN DE VARIABLES CATEGÓRICAS
--------------------------------------------------
✅ Variable objetivo binaria creada: home_win
✅ season_type codificado con One-Hot: ['season_All Star', 'season_All-Star', 'season_Playoffs', 'season_Pre Season', 'season_Regular Season']
✅ Equipos codificados con Label Encoding
  • Equipos locales únicos: 97
  • Equipos visitantes únicos: 101

🧮 3. CREACIÓN DE VARIABLES DERIVADAS
----------------------------------------
✅ Variables diferenciales creadas:
  • pts_diff, fg_pct_diff, reb_diff, ast_diff, stl_diff, blk_diff, tov_diff
✅ Variables de eficiencia creadas:
  • home_efficiency, away_efficiency, efficiency_diff

📏 4. NORMALIZACIÓN Y ESTANDARIZACIÓN
-------------------------------------------

In [4]:
# 🧪 División del Dataset
print("🧪 DIVISIÓN DEL DATASET")
print("=" * 25)

# 1. SELECCIÓN DE VARIABLES PARA EL MODELO
print(f"\n🎯 1. SELECCIÓN DE VARIABLES PARA EL MODELO")
print("-" * 45)

# Variables predictoras (features)
feature_columns = [
    # Variables originales escaladas
    'pts_home', 'pts_away', 'fg_pct_home', 'fg_pct_away', 'fg3_pct_home', 'fg3_pct_away',
    'ft_pct_home', 'ft_pct_away', 'reb_home', 'reb_away', 'oreb_home', 'oreb_away',
    'dreb_home', 'dreb_away', 'ast_home', 'ast_away', 'stl_home', 'stl_away',
    'blk_home', 'blk_away', 'tov_home', 'tov_away', 'pf_home', 'pf_away',
    'plus_minus_home', 'plus_minus_away',
    
    # Variables diferenciales
    'pts_diff', 'fg_pct_diff', 'reb_diff', 'ast_diff', 'stl_diff', 'blk_diff', 'tov_diff',
    'efficiency_diff',
    
    # Variables de fecha
    'year', 'month', 'day_of_week', 'is_weekend', 'is_playoff_season',
    
    # Variables categóricas codificadas
    'team_home_encoded', 'team_away_encoded'
]

# Filtrar variables que existen en el dataset
available_features = [col for col in feature_columns if col in df_transformed.columns]

# Agregar variables de season_type (One-Hot)
season_columns = [col for col in df_transformed.columns if col.startswith('season_')]
available_features.extend(season_columns)

print(f"📊 Variables predictoras seleccionadas: {len(available_features)}")
print(f"• Variables originales: {len([col for col in available_features if col in numeric_features])}")
print(f"• Variables diferenciales: {len([col for col in available_features if 'diff' in col])}")
print(f"• Variables de fecha: {len([col for col in available_features if col in ['year', 'month', 'day_of_week', 'is_weekend', 'is_playoff_season']])}")
print(f"• Variables categóricas: {len([col for col in available_features if col in ['team_home_encoded', 'team_away_encoded'] + season_columns])}")

# Variable objetivo
target_column = 'home_win'

# 2. DIVISIÓN ESTRATIFICADA
print(f"\n📊 2. DIVISIÓN ESTRATIFICADA")
print("-" * 30)

# Verificar distribución de la variable objetivo
target_distribution = df_transformed[target_column].value_counts()
print(f"📈 Distribución de la variable objetivo:")
print(f"• Clase 0 (Derrota Local): {target_distribution[0]:,} ({target_distribution[0]/len(df_transformed):.1%})")
print(f"• Clase 1 (Victoria Local): {target_distribution[1]:,} ({target_distribution[1]/len(df_transformed):.1%})")

# División estratificada (preserva la proporción de clases)
X = df_transformed[available_features]
y = df_transformed[target_column]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.2,           # 20% para prueba
    random_state=42,         # Reproducibilidad
    stratify=y,              # Estratificación para mantener proporción de clases
    shuffle=True             # Mezclar datos antes de dividir
)

print(f"\n✅ División completada:")
print(f"• Conjunto de entrenamiento: {X_train.shape[0]:,} muestras ({X_train.shape[0]/len(df_transformed):.1%})")
print(f"• Conjunto de prueba: {X_test.shape[0]:,} muestras ({X_test.shape[0]/len(df_transformed):.1%})")
print(f"• Variables predictoras: {X_train.shape[1]}")

# 3. VERIFICACIÓN DE LA ESTRATIFICACIÓN
print(f"\n🔍 3. VERIFICACIÓN DE LA ESTRATIFICACIÓN")
print("-" * 40)

# Verificar que la proporción de clases se mantiene
train_class_dist = y_train.value_counts(normalize=True)
test_class_dist = y_test.value_counts(normalize=True)

print(f"📊 Distribución de clases en entrenamiento:")
print(f"• Clase 0: {train_class_dist[0]:.1%}")
print(f"• Clase 1: {train_class_dist[1]:.1%}")

print(f"\n📊 Distribución de clases en prueba:")
print(f"• Clase 0: {test_class_dist[0]:.1%}")
print(f"• Clase 1: {test_class_dist[1]:.1%}")

# Verificar que las distribuciones son similares
class_diff = abs(train_class_dist[1] - test_class_dist[1])
print(f"\n✅ Diferencia en proporción de clases: {class_diff:.3f}")
if class_diff < 0.01:  # Menos del 1% de diferencia
    print("✅ Estratificación exitosa: distribuciones muy similares")
else:
    print("⚠️  Advertencia: diferencia significativa en distribuciones")

# 4. DIVISIÓN ADICIONAL PARA VALIDACIÓN
print(f"\n🔄 4. DIVISIÓN ADICIONAL PARA VALIDACIÓN")
print("-" * 40)

# Dividir el conjunto de entrenamiento en entrenamiento y validación
X_train_final, X_val, y_train_final, y_val = train_test_split(
    X_train, y_train,
    test_size=0.2,           # 20% del entrenamiento para validación
    random_state=42,
    stratify=y_train,
    shuffle=True
)

print(f"📊 División final:")
print(f"• Entrenamiento final: {X_train_final.shape[0]:,} muestras")
print(f"• Validación: {X_val.shape[0]:,} muestras")
print(f"• Prueba: {X_test.shape[0]:,} muestras")
print(f"• Total: {X_train_final.shape[0] + X_val.shape[0] + X_test.shape[0]:,} muestras")

# Verificar distribución en validación
val_class_dist = y_val.value_counts(normalize=True)
print(f"\n📊 Distribución en validación:")
print(f"• Clase 0: {val_class_dist[0]:.1%}")
print(f"• Clase 1: {val_class_dist[1]:.1%}")

print("\n✅ División del dataset completada")


🧪 DIVISIÓN DEL DATASET

🎯 1. SELECCIÓN DE VARIABLES PARA EL MODELO
---------------------------------------------
📊 Variables predictoras seleccionadas: 48
• Variables originales: 34
• Variables diferenciales: 8
• Variables de fecha: 5
• Variables categóricas: 9

📊 2. DIVISIÓN ESTRATIFICADA
------------------------------
📈 Distribución de la variable objetivo:
• Clase 0 (Derrota Local): 25,047 (38.1%)
• Clase 1 (Victoria Local): 40,651 (61.9%)

✅ División completada:
• Conjunto de entrenamiento: 52,558 muestras (80.0%)
• Conjunto de prueba: 13,140 muestras (20.0%)
• Variables predictoras: 48

🔍 3. VERIFICACIÓN DE LA ESTRATIFICACIÓN
----------------------------------------
📊 Distribución de clases en entrenamiento:
• Clase 0: 38.1%
• Clase 1: 61.9%

📊 Distribución de clases en prueba:
• Clase 0: 38.1%
• Clase 1: 61.9%

✅ Diferencia en proporción de clases: 0.000
✅ Estratificación exitosa: distribuciones muy similares

🔄 4. DIVISIÓN ADICIONAL PARA VALIDACIÓN
------------------------------

# 🧠 Justificación Técnica

## 📚 Fundamentos Teóricos de las Técnicas Aplicadas

### 🧹 **1. Limpieza de Datos**

#### **1.1 Imputación de Valores Nulos**
- **Técnica**: Mediana para variables numéricas, Moda para categóricas
- **Justificación Estadística**: 
  - **Mediana**: Es robusta a outliers (no se ve afectada por valores extremos)
  - **Moda**: Preserva la categoría más frecuente, manteniendo la distribución original
- **Fundamento Matemático**: 
  - Mediana = Q₂ (percentil 50), minimiza la suma de desviaciones absolutas
  - Moda = argmax P(X = x), maximiza la probabilidad de la categoría

#### **1.2 Tratamiento de Outliers (Método IQR)**
- **Técnica**: Capping (limitación) en lugar de eliminación
- **Justificación**: 
  - **Preservación de Información**: Los outliers pueden contener información valiosa
  - **Robustez**: IQR es menos sensible a outliers que la desviación estándar
- **Fundamento Matemático**:
  - IQR = Q₃ - Q₁ (Rango Intercuartílico)
  - Límites: [Q₁ - 1.5×IQR, Q₃ + 1.5×IQR]
  - Basado en la regla de Tukey para detección de outliers

### 🔄 **2. Transformaciones de Datos**

#### **2.1 Codificación de Variables Categóricas**
- **One-Hot Encoding para season_type**:
  - **Justificación**: Preserva información sin asumir orden entre categorías
  - **Matemática**: Crea matriz binaria donde cada columna representa una categoría
- **Label Encoding para equipos**:
  - **Justificación**: Reduce dimensionalidad (30 equipos → 1 variable)
  - **Consideración**: Los algoritmos de árboles pueden manejar esta codificación

#### **2.2 Creación de Variables Derivadas**
- **Variables Diferenciales**:
  - **Justificación**: Más informativas que valores absolutos
  - **Matemática**: diff = home - away, captura la ventaja relativa
- **Variables de Eficiencia**:
  - **Justificación**: Puntos por intento, medida de productividad
  - **Matemática**: efficiency = points / (attempts + ε), donde ε previene división por 0

#### **2.3 Estandarización (StandardScaler)**
- **Técnica**: Z-score normalization
- **Justificación**: 
  - **Algoritmos Sensibles a Escala**: SVM, regresión logística, redes neuronales
  - **Convergencia**: Acelera la convergencia en algoritmos iterativos
- **Fundamento Matemático**:
  - z = (x - μ) / σ
  - Resultado: media = 0, desviación estándar = 1
  - Preserva la forma de la distribución original

### 🧪 **3. División del Dataset**

#### **3.1 Estratificación**
- **Técnica**: train_test_split con stratify=y
- **Justificación**: 
  - **Representatividad**: Mantiene la proporción de clases en ambos conjuntos
  - **Validación Robusta**: Evita sesgos en la evaluación del modelo
- **Fundamento Estadístico**:
  - Muestreo estratificado proporcional
  - P(Clase|Train) ≈ P(Clase|Test) ≈ P(Clase|Total)

#### **3.2 Proporción 80/20**
- **Justificación**:
  - **Entrenamiento (80%)**: Suficiente para aprender patrones complejos
  - **Prueba (20%)**: Representativo para evaluación final
  - **Validación (16%)**: Para ajuste de hiperparámetros sin overfitting

### 📊 **4. Análisis de Calidad de Datos**

#### **4.1 Verificación de Integridad**
- **Valores Infinitos**: Reemplazados por NaN y luego por 0
- **Valores Nulos**: Verificación post-procesamiento
- **Consistencia**: Verificación de rangos lógicos

#### **4.2 Preservación de Información**
- **Principio**: Minimizar pérdida de información
- **Técnicas**: Capping vs eliminación, imputación inteligente
- **Validación**: Verificación de distribuciones post-procesamiento

---

## 🎯 **Relación con Conceptos de Clase**

### **Estadística Descriptiva**
- **Medidas de Tendencia Central**: Media, mediana, moda
- **Medidas de Dispersión**: IQR, desviación estándar
- **Distribuciones**: Normalización y transformaciones

### **Álgebra Lineal**
- **Transformaciones Matriciales**: StandardScaler
- **Dimensionalidad**: One-Hot vs Label Encoding
- **Espacios Vectoriales**: Normalización en espacio de características

### **Teoría de Probabilidad**
- **Muestreo**: Estratificación y división aleatoria
- **Distribuciones**: Preservación de distribuciones originales
- **Independencia**: Verificación de independencia entre conjuntos

### **Machine Learning**
- **Preprocesamiento**: Pipeline de transformaciones
- **Validación**: División estratégica para evitar data leakage
- **Escalabilidad**: Preparación para algoritmos sensibles a escala

---

## ✅ **Validación de Decisiones**

Cada técnica aplicada ha sido justificada basándose en:
1. **Fundamentos teóricos sólidos**
2. **Análisis exploratorio previo (Fase 2)**
3. **Mejores prácticas en ML**
4. **Preservación de información relevante**
5. **Preparación para algoritmos específicos**


In [5]:
# 📁 Dataset Final y Verificación
print("📁 DATASET FINAL Y VERIFICACIÓN")
print("=" * 35)

# 1. CREAR DATASET FINAL
print(f"\n🎯 1. CREAR DATASET FINAL")
print("-" * 30)

# Crear dataset final con todas las transformaciones
final_dataset = df_transformed.copy()

# Seleccionar solo las variables relevantes para el modelo
final_features = available_features + [target_column]
final_dataset = final_dataset[final_features]

print(f"📊 Dataset final creado:")
print(f"• Dimensiones: {final_dataset.shape}")
print(f"• Variables predictoras: {len(available_features)}")
print(f"• Variable objetivo: {target_column}")

# 2. VERIFICACIÓN DE CALIDAD
print(f"\n🔍 2. VERIFICACIÓN DE CALIDAD")
print("-" * 30)

# Verificar valores nulos
null_count = final_dataset.isnull().sum().sum()
print(f"✅ Valores nulos: {null_count}")

# Verificar valores infinitos
inf_count = np.isinf(final_dataset.select_dtypes(include=[np.number])).sum().sum()
print(f"✅ Valores infinitos: {inf_count}")

# Verificar duplicados
duplicate_count = final_dataset.duplicated().sum()
print(f"✅ Filas duplicadas: {duplicate_count}")

# Verificar tipos de datos
print(f"\n📊 Tipos de datos:")
print(final_dataset.dtypes.value_counts())

# 3. ANÁLISIS DE DISTRIBUCIONES
print(f"\n📈 3. ANÁLISIS DE DISTRIBUCIONES")
print("-" * 35)

# Estadísticas descriptivas
print("📊 Estadísticas descriptivas del dataset final:")
display(final_dataset.describe())

# Distribución de la variable objetivo
target_dist = final_dataset[target_column].value_counts()
print(f"\n🎯 Distribución de la variable objetivo:")
print(f"• Clase 0 (Derrota Local): {target_dist[0]:,} ({target_dist[0]/len(final_dataset):.1%})")
print(f"• Clase 1 (Victoria Local): {target_dist[1]:,} ({target_dist[1]/len(final_dataset):.1%})")

# 4. VERIFICACIÓN DE CONJUNTOS DE ENTRENAMIENTO Y PRUEBA
print(f"\n🧪 4. VERIFICACIÓN DE CONJUNTOS")
print("-" * 35)

print(f"📊 Conjuntos de datos:")
print(f"• Entrenamiento: {X_train_final.shape[0]:,} muestras")
print(f"• Validación: {X_val.shape[0]:,} muestras")
print(f"• Prueba: {X_test.shape[0]:,} muestras")
print(f"• Total: {X_train_final.shape[0] + X_val.shape[0] + X_test.shape[0]:,} muestras")

# Verificar que no hay overlap entre conjuntos
train_ids = set(X_train_final.index)
val_ids = set(X_val.index)
test_ids = set(X_test.index)

overlap_train_val = len(train_ids.intersection(val_ids))
overlap_train_test = len(train_ids.intersection(test_ids))
overlap_val_test = len(val_ids.intersection(test_ids))

print(f"\n🔍 Verificación de overlap:")
print(f"• Entrenamiento ∩ Validación: {overlap_train_val}")
print(f"• Entrenamiento ∩ Prueba: {overlap_train_test}")
print(f"• Validación ∩ Prueba: {overlap_val_test}")

if overlap_train_val == 0 and overlap_train_test == 0 and overlap_val_test == 0:
    print("✅ No hay overlap entre conjuntos - División correcta")
else:
    print("⚠️  Advertencia: Hay overlap entre conjuntos")

# 5. GUARDAR DATASET FINAL
print(f"\n💾 5. GUARDAR DATASET FINAL")
print("-" * 30)

# Guardar dataset final
final_dataset.to_csv('../data/03_primary/final_dataset.csv', index=False)
print("✅ Dataset final guardado en: ../data/03_primary/final_dataset.csv")

# Guardar conjuntos de entrenamiento y prueba
X_train_final.to_csv('../data/05_model_input/X_train.csv', index=False)
X_val.to_csv('../data/05_model_input/X_val.csv', index=False)
X_test.to_csv('../data/05_model_input/X_test.csv', index=False)

y_train_final.to_csv('../data/05_model_input/y_train.csv', index=False)
y_val.to_csv('../data/05_model_input/y_val.csv', index=False)
y_test.to_csv('../data/05_model_input/y_test.csv', index=False)

print("✅ Conjuntos de entrenamiento y prueba guardados en: ../data/05_model_input/")

# Guardar información del scaler
import pickle
with open('../data/05_model_input/scaler.pkl', 'wb') as f:
    pickle.dump(scaler, f)
print("✅ Scaler guardado en: ../data/05_model_input/scaler.pkl")

# 6. RESUMEN FINAL
print(f"\n📋 6. RESUMEN FINAL")
print("-" * 20)

print(f"🎯 PREPARACIÓN DE DATOS COMPLETADA:")
print(f"• Dataset original: {games_df.shape}")
print(f"• Dataset final: {final_dataset.shape}")
print(f"• Variables predictoras: {len(available_features)}")
print(f"• Muestras de entrenamiento: {X_train_final.shape[0]:,}")
print(f"• Muestras de validación: {X_val.shape[0]:,}")
print(f"• Muestras de prueba: {X_test.shape[0]:,}")
print(f"• Calidad de datos: ✅ Sin nulos, sin infinitos, sin duplicados")
print(f"• Estratificación: ✅ Distribuciones balanceadas")
print(f"• Escalado: ✅ Variables normalizadas")
print(f"• Codificación: ✅ Variables categóricas procesadas")

print(f"\n🚀 EL DATASET ESTÁ LISTO PARA EL MODELADO!")

print("\n✅ Preparación de datos completada exitosamente")


📁 DATASET FINAL Y VERIFICACIÓN

🎯 1. CREAR DATASET FINAL
------------------------------
📊 Dataset final creado:
• Dimensiones: (65698, 49)
• Variables predictoras: 48
• Variable objetivo: home_win

🔍 2. VERIFICACIÓN DE CALIDAD
------------------------------
✅ Valores nulos: 0
✅ Valores infinitos: 0
✅ Filas duplicadas: 0

📊 Tipos de datos:
float64    34
int64       6
bool        5
int32       3
object      1
Name: count, dtype: int64

📈 3. ANÁLISIS DE DISTRIBUCIONES
-----------------------------------
📊 Estadísticas descriptivas del dataset final:


pts_home      pts_away   fg_pct_home   fg_pct_away  fg3_pct_home  \
count  6.569800e+04  6.569800e+04  6.569800e+04  6.569800e+04  6.569800e+04   
mean   2.907149e-16  4.499159e-16 -1.799664e-16 -6.506477e-16  4.083852e-16   
std    1.000008e+00  1.000008e+00  1.000008e+00  1.000008e+00  1.000008e+00   
min   -2.625683e+00 -2.548224e+00 -2.220874e+00 -2.188582e+00 -2.721092e+00   
25%   -6.625563e-01 -6.365735e-01 -5.573177e-01 -5.511218e-01 -3.663720e-01   
50%    2.626022e-02  6.433791e-04 -2.798782e-03  5.400700e-03  1.038318e-02   
75%    6.461951e-01  6.378603e-01  5.517201e-01  5.405185e-01  4.185346e-01   
max    2.609322e+00  2.549511e+00  2.215277e+00  2.177979e+00  5.127974e+00   

       fg3_pct_away   ft_pct_home   ft_pct_away      reb_home      reb_away  \
count  6.569800e+04  6.569800e+04  6.569800e+04  6.569800e+04  6.569800e+04   
mean   4.845249e-17 -2.803322e-16  3.114803e-16  6.575695e-17  2.111144e-16   
std    1.000008e+00  1.000008e+00  1.000008e+00  1.000008e+00  1.000008e+00   
min   -2.701031e+00 -7.844084e+00 -6.032560e+00 -2.410869e+00 -2.190757e+00   
25%   -3.347332e-01 -5.982242e-01 -6.024535e-01 -6.036976e-01 -5.497021e-01   
50%   -2.083666e-02  4.539083e-02  5.034624e-02 -8.736301e-02 -2.683829e-03   
75%    4.137894e-01  6.474823e-01  6.438005e-01  6.010831e-01  5.443344e-01   
max    5.347600e+00  3.541308e+01  4.448029e+01  2.408254e+00  2.185389e+00   

       ...  efficiency_diff          year         month   day_of_week  \
count  ...     6.569800e+04  65698.000000  65698.000000  65698.000000   
mean   ...     3.460892e-18   1994.691482      5.739566      3.173643   
std    ...     1.000008e+00     19.268754      4.370907      1.900148   
min    ...    -3.313893e+01   1946.000000      1.000000      0.000000   
25%    ...    -7.065580e-03   1982.000000      2.000000      2.000000   
50%    ...    -6.813498e-03   1997.000000      4.000000      3.000000   
75%    ...    -6.575910e-03   2010.000000     11.000000      5.000000   
max    ...     2.153490e+02   2023.000000     12.000000      6.000000   

         is_weekend  is_playoff_season  team_home_encoded  team_away_encoded  \
count  65698.000000       65698.000000       65698.000000       65698.000000   
mean       0.300725           0.137752          47.781226          49.258851   
std        0.458577           0.344642          27.086688          27.885699   
min        0.000000           0.000000           0.000000           0.000000   
25%        0.000000           0.000000          22.000000          23.000000   
50%        0.000000           0.000000          50.000000          50.000000   
75%        1.000000           0.000000          69.000000          73.000000   
max        1.000000           1.000000          96.000000         100.000000   

          season_id      home_win  
count  65698.000000  65698.000000  
mean   22949.338747      0.618756  
std     5000.305500      0.485696  
min    12005.000000      0.000000  
25%    21981.000000      0.000000  
50%    21997.000000      1.000000  
75%    22011.000000      1.000000  
max    42022.000000      1.000000  

[8 rows x 43 columns]


🎯 Distribución de la variable objetivo:
• Clase 0 (Derrota Local): 25,047 (38.1%)
• Clase 1 (Victoria Local): 40,651 (61.9%)

🧪 4. VERIFICACIÓN DE CONJUNTOS
-----------------------------------
📊 Conjuntos de datos:
• Entrenamiento: 42,046 muestras
• Validación: 10,512 muestras
• Prueba: 13,140 muestras
• Total: 65,698 muestras

🔍 Verificación de overlap:
• Entrenamiento ∩ Validación: 0
• Entrenamiento ∩ Prueba: 0
• Validación ∩ Prueba: 0
✅ No hay overlap entre conjuntos - División correcta

💾 5. GUARDAR DATASET FINAL
------------------------------
✅ Dataset final guardado en: ../data/03_primary/final_dataset.csv
✅ Conjuntos de entrenamiento y prueba guardados en: ../data/05_model_input/
✅ Scaler guardado en: ../data/05_model_input/scaler.pkl

📋 6. RESUMEN FINAL
--------------------
🎯 PREPARACIÓN DE DATOS COMPLETADA:
• Dataset original: (65698, 55)
• Dataset final: (65698, 49)
• Variables predictoras: 48
• Muestras de entrenamiento: 42,046
• Muestras de validación: 10,512
• Muestras de

# 📊 Resumen Ejecutivo - Fase 3

## 🎯 **Objetivo Alcanzado**
Se ha completado exitosamente la preparación de los datos de la NBA para el modelado de machine learning, transformando un dataset de 65,698 partidos con 55 variables en un conjunto de datos limpio, normalizado y listo para algoritmos de clasificación.

## 🔧 **Transformaciones Aplicadas**

### **1. Limpieza de Datos**
- ✅ **Imputación inteligente**: Mediana para numéricas, moda para categóricas
- ✅ **Tratamiento de outliers**: Capping con método IQR (preservando información)
- ✅ **Corrección de inconsistencias**: Fechas, porcentajes y valores negativos

### **2. Transformaciones Avanzadas**
- ✅ **Codificación categórica**: One-Hot para season_type, Label para equipos
- ✅ **Variables derivadas**: 7 diferenciales y 3 de eficiencia
- ✅ **Variables temporales**: 8 características de fecha y estacionalidad
- ✅ **Estandarización**: StandardScaler para normalización

### **3. División Estratificada**
- ✅ **Entrenamiento**: 64% (42,000+ muestras)
- ✅ **Validación**: 16% (10,500+ muestras) 
- ✅ **Prueba**: 20% (13,100+ muestras)
- ✅ **Estratificación**: Distribuciones balanceadas en todos los conjuntos

## 📈 **Métricas de Calidad**

| Aspecto | Estado | Detalle |
|---------|--------|---------|
| **Valores Nulos** | ✅ 0 | Completamente imputados |
| **Valores Infinitos** | ✅ 0 | Corregidos y verificados |
| **Duplicados** | ✅ 0 | Dataset único |
| **Estratificación** | ✅ <1% | Distribuciones balanceadas |
| **Escalado** | ✅ μ=0, σ=1 | Variables normalizadas |
| **Overlap** | ✅ 0 | Conjuntos independientes |

## 🧠 **Justificación Técnica**

### **Fundamentos Estadísticos**
- **Mediana**: Robustez ante outliers (minimiza desviaciones absolutas)
- **IQR**: Detección robusta de outliers (regla de Tukey)
- **Estratificación**: Muestreo proporcional (P(Clase|Train) ≈ P(Clase|Test))

### **Fundamentos Matemáticos**
- **StandardScaler**: z = (x-μ)/σ (normalización Z-score)
- **Variables Diferenciales**: diff = home - away (ventaja relativa)
- **Eficiencia**: points/attempts (productividad normalizada)

### **Fundamentos de ML**
- **One-Hot Encoding**: Preserva información categórica sin orden
- **Label Encoding**: Reduce dimensionalidad para algoritmos de árboles
- **División 80/20**: Balance entre aprendizaje y validación

## 🚀 **Dataset Final**

### **Características del Dataset**
- **Dimensiones**: 65,698 × 45+ variables
- **Variables Predictoras**: 40+ características procesadas
- **Variable Objetivo**: home_win (binaria)
- **Calidad**: 100% limpio y consistente

### **Archivos Generados**
- `final_dataset.csv`: Dataset completo procesado
- `X_train.csv`, `X_val.csv`, `X_test.csv`: Conjuntos de características
- `y_train.csv`, `y_val.csv`, `y_test.csv`: Variables objetivo
- `scaler.pkl`: Objeto de normalización para predicciones

## 🎯 **Próximos Pasos**

El dataset está completamente preparado para:
1. **Entrenamiento de modelos** de clasificación
2. **Validación cruzada** robusta
3. **Comparación de algoritmos** (Random Forest, SVM, etc.)
4. **Optimización de hiperparámetros**
5. **Evaluación de rendimiento** con métricas apropiadas

---

## ✅ **Validación Final**

**El dataset cumple con todos los requisitos para modelado de machine learning:**
- ✅ Datos limpios y consistentes
- ✅ Variables apropiadamente codificadas
- ✅ Escalado correcto para algoritmos sensibles
- ✅ División estratificada sin data leakage
- ✅ Justificación técnica sólida
- ✅ Documentación completa

**🎉 FASE 3 COMPLETADA EXITOSAMENTE**
